create-hpc-cluster/shared-resources/mainTemplate.json

```diff
@@ -0,0 +1,639 @@
+    "privateClusterFQDN": "[concat(parameters('clusterName'), '.', parameters('domainName'))]",
+    "vmssSinglePlacementGroup": "[lessOrEquals(parameters('computeNodeNumber'), if(variables('computeNodeUsesCustomImage'), 40, 100))]",
+    "vnetRGName": "[if(equals(parameters('newOrExistingDomainForest'), 'new'), resourceGroup().name, parameters('virtualNetworkResourceGroupName'))]",
+    "addressPrefix": "10.0.0.0/16",
+    "subnetAddressPrefix": "10.0.0.0/22",
+    "hnPrivateIPAllocationMethod": "[if(equals(parameters('newOrExistingDomainForest'), 'new'), 'Static', 'Dynamic')]",
+    "headNoePrivateIP": "[if(equals(parameters('newOrExistingDomainForest'), 'new'), '10.0.0.4', '')]",
+    "vnetID": "[resourceId(variables('vnetRGName'), 'Microsoft.Network/virtualNetworks',parameters('virtualNetworkName'))]",
+    "subnetRef": "[concat(variables('vnetID'),'/subnets/', parameters('subnetName'))]",
+    "defaultPublicIPName": "[concat(parameters('clusterName'), '-PublicIP-HN')]",
+    "publicIPAddressName": "[if(and(empty(parameters('publicIPName')), equals(parameters('publicIPCreateOption'), 'new')), variables('defaultPublicIPName'), parameters('publicIPName'))]",
+    "publicIPDNSNameLabel": "[if(and(empty(parameters('publicIPDNSNameLabel')), equals(parameters('publicIPCreateOption'), 'new')), parameters('clusterName'), parameters('publicIPDNSNameLabel'))]",
+    "availabilitySetNamePrefix": "[concat(parameters('clusterName'), '-avset')]",
+    "nbrVMPerAvailabilitySet": 180,
+    "availabilitySetNumber": "[add(div(parameters('computeNodeNumber'), variables('nbrVMPerAvailabilitySet')), 1)]",
+    "nicNameSuffix": "[concat('-nic', uniqueString(variables('subnetRef')))]",
+    "nicNameHN": "[concat(parameters('clusterName'), variables('nicNameSuffix'))]",
+    "rdmaVMSizes": [
+      "Standard_A8",
+      "Standard_A9",
+      "Standard_H16r",
+      "Standard_H16mr",
+      "Standard_NC24r"
+    ],
+    "cnInstallRDMADriver": "[contains(variables('rdmaVMSizes'), parameters('computeNodeVMSize'))]",
+    "hnInstallRDMADriver": "[contains(variables('rdmaVMSizes'), parameters('headNodeVMSize'))]",
+    "hnAvailabilitySetName": "[if(and(and(variables('cnInstallRDMADriver'), variables('hnInstallRDMADriver')), not(parameters('useVmssForComputeNodes'))), concat(variables('availabilitySetNamePrefix'), '00'), '')]",
+    "computeVmssName": "[take(replace(parameters('computeNodeNamePrefix'), '-', ''), 9)]",
+    "computeNodeTemplateName": "[if(equals(parameters('computeNodeImageOsPlatform'), 'Linux'), 'linuxnode.json', 'windowsnode.json')]",
+    "computeNodeTemplateUrl": "[concat(parameters('artifactsBaseUrl'), '/', variables('computeNodeTemplateName'))]",
+    "computeNodeVmssTemplateName": "[if(equals(parameters('computeNodeImageOsPlatform'), 'Linux'), 'linuxvmss.json', 'windowsvmss.json')]",
+    "computeNodeVmssTemplateUrl": "[concat(parameters('artifactsBaseUrl'), '/', variables('computeNodeVmssTemplateName'))]",
+    "fakeCusomImageReference": {
```

Comment:
No, this is a trick, we want to record the "offer" and "sku" of the compute node image in the customdata for official images, but for the custom image, there is only "id" field in the imagereference, so we make a fake image reference for custom image, so that when we fetch the "offer"/"sku" field it will not fail.

Reply:
The customData property expects bogus data?  Also, I don't see a path through this sample that would provide an id property on the image object - it's either with Excel or without...

Reply:
We use the customData to pass image information to the VM, so that we can track the image usage. In another template create-hpc-cluster-custom-image, a custom image will be used. we have 3 templates here, they shares most nested templates, so i put all the shared resources in one template create-hpc-cluster, and the other two will also use the shared resources in create-hpc-cluster template folder.